### ISSUE TITLE
No common parent subvolume found

### ISSUE THREAD

**Issue by Massimo-B** [digint/btrbk#182]
Because I wondered that some transfers are taking very long I enabled incremental strict.
Now sometimes no common parent is found on the target, but there is one:
```bash
# ls -ald /mnt/usb/mobiledata/snapshots/home/home.20170911T071634+0200/
drwxr-xr-x 1 root root 78 Sep 11 07:24 /mnt/usb/mobiledata/snapshots/home/home.20170911T071634+0200//
# ls -ald /.snapshots/home/home.20170911T071634+0200/
drwxr-xr-x 1 root root 78 Aug 17  2016 /.snapshots/home/home.20170911T071634+0200//
```

Looking at the UUIDs the id on the target is different:
```bash
# btrfs subvolume show /mnt/usb/mobiledata/snapshots/home/home.20170911T071634+0200/
snapshots/home/home.20170911T071634+0200
	Name: 			home.20170911T071634+0200
	UUID: 			c873c0a2-1388-9d44-9e43-63573d7d9a35
	Parent UUID: 		1ed92455-df64-cf44-9f77-06f78c7f8fa7
	Received UUID: 		26105b96-f945-1442-9a17-72b1fc524830
	Creation time: 		2017-09-11 07:23:57 +0200
	Subvolume ID: 		1932
	Generation: 		5193
	Gen at creation: 	5189
	Parent ID: 		257
	Top level ID: 		257
	Flags: 			readonly
	Snapshot(s):
# btrfs subvolume show /.snapshots/home/home.20170911T071634+0200/
home/home.20170911T071634+0200
	Name: 			home.20170911T071634+0200
	UUID: 			26105b96-f945-1442-9a17-72b1fc524830
	Parent UUID: 		c593b1dd-973a-fe4d-9f1f-68b0c7e2b872
	Received UUID: 		-
	Creation time: 		2017-09-11 07:16:42 +0200
	Subvolume ID: 		6521
	Generation: 		101555
	Gen at creation: 	101554
	Parent ID: 		261
	Top level ID: 		261
	Flags: 			readonly
	Snapshot(s):
```
How can the uuid have changed eventhough the snapshot name is equal and why isn't that used as common parent as it exists on both btrfs?

```bash
# btrbk --progress run
Creating backup: /mnt/usb/data/snapshots/root/root.20170915T073515+0200
 292KiB 0:00:06 [43,8KiB/s] [43,8KiB/s]
WARNING: Backup to /mnt/usb/mobiledata/snapshots/root failed: no common parent subvolume found for "/mnt/btrfs-top-lvl/snapshots/root/root.20170807T080103+0200", and option "incremental" is set to "strict"
ERROR: Error while resuming backups, aborting
Creating backup: /mnt/usb/data/snapshots/home/home.20170915T073515+0200
10,4MiB 0:00:12 [ 866KiB/s] [ 866KiB/s]
WARNING: Backup to /mnt/usb/mobiledata/snapshots/home failed: no common parent subvolume found for "/mnt/btrfs-top-lvl/snapshots/home/home.20170607T071118+0200", and option "incremental" is set to "strict"
ERROR: Error while resuming backups, aborting
WARNING: Skipping cleanup of snapshots for subvolume "/mnt/btrfs-top-lvl/root", as at least one target aborted earlier
WARNING: Skipping cleanup of snapshots for subvolume "/mnt/btrfs-top-lvl/home", as at least one target aborted earlier
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.25.1)

    Date:   Fri Sep 15 07:35:15 2017
    Config: /etc/btrbk/btrbk.conf

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/mnt/btrfs-top-lvl/root
+++ /mnt/btrfs-top-lvl/snapshots/root/root.20170915T073515+0200
>>> /mnt/usb/data/snapshots/root/root.20170915T073515+0200
--- /mnt/usb/data/snapshots/root/root.20170915T071720+0200
!!! Target "/mnt/usb/mobiledata/snapshots/root" aborted: No common parent subvolume found, and option "incremental" is set to "strict"

/mnt/btrfs-top-lvl/home
+++ /mnt/btrfs-top-lvl/snapshots/home/home.20170915T073515+0200
>>> /mnt/usb/data/snapshots/home/home.20170915T073515+0200
!!! Target "/mnt/usb/mobiledata/snapshots/home" aborted: No common parent subvolume found, and option "incremental" is set to "strict"

/mnt/btrfs-top-lvl/data
=== /mnt/btrfs-top-lvl/snapshots/data/data.20170915T063607+0200

NOTE: Some errors occurred, which may result in missing backups!
Please check warning and error messages above.
```
```bash
# cat /etc/btrbk/btrbk.conf
transaction_log         /var/log/btrbk.log
timestamp_format        long-iso

incremental             strict
archive_preserve_min    latest
archive_preserve        12m 10y

volume /mnt/btrfs-top-lvl/
    snapshot_create  onchange

    subvolume root
        group root
        snapshot_dir            snapshots/root
        snapshot_preserve_min   5d
        snapshot_preserve       2d 1w 1m
        target send-receive     /mnt/usb/data/snapshots/root
            target_preserve_min     latest
            target_preserve         1d 4w 2m 1y
        target send-receive     /mnt/usb/mobiledata/snapshots/root
            target_preserve_min     latest
            target_preserve         1d 4w 2m 1y
    
    subvolume home
        group home
        snapshot_dir            snapshots/home
        snapshot_preserve_min   3d
        snapshot_preserve       4d 4w 2m 1y
        target send-receive     /mnt/usb/data/snapshots/home
            target_preserve_min     3d
            target_preserve         6d 5w 4m 3y
        target send-receive     /mnt/usb/mobiledata/snapshots/home
            target_preserve_min     3d
            target_preserve         6d 5w 4m 3y
        
    subvolume data
        group data
        snapshot_dir            snapshots/data
        snapshot_preserve_min   latest
        snapshot_preserve       2w
        #target send-receive     /mnt/usb/data/snapshots/data
            #target_preserve_min     latest
            #target_preserve         2m
        #target send-receive     /mnt/usb/mobiledata/snapshots/data
            #target_preserve_min     latest
            #target_preserve         2m
```

**Comment by digint** [comment]
> How can the uuid have changed eventhough the snapshot name is equal and why isn't that used as common parent as it exists on both btrfs?

It's a different snapshot, so it has different UUID. From user perspective, the name is just a directory name, and can be changed (e.g. with "mv").

Even more, the snapshots were made from different parents (do you have several "home" subvolumes?). From your config I can't see who generated the snapshot in `/.snapshots`.

> Because I wondered that some transfers are taking very long I enabled incremental strict.
Now sometimes no common parent is found on the target, but there is one:

Probably there's none... a common parent for send-receive is determined by matching uuid <--> received_uuid (NOT parent_uuid, this only tells btrfs from which subvolume the snapshot was created).

**Comment by Massimo-B** [comment]
Please add the mentioned feature of "**recursive deep-resolving**".
This is very important as I have many full transfers when incremental would be possible by just taking the next parent.

As you argued that this actually would never happen as btrbk keeps local snapshots as long as until the last is transferred to the last missing target. But I guess if retention is applied first, snapshots are getting deleted and missing if the target is available for the next time.
I mean you could still add some debug logs to log when and why that recursive deep-resolving has been started out there, I can provide log files of that test version back to you.

**Comment by digint** [comment]
Recursive search on parent_uuid chain is implemented in: 3be65b9f67ff71ac215a5006740b4bf1fdd8b98c
